4658:add custom model rules resource

matt-metivier opened the pull request:
This PR introduces a new Terraform resource for managing Asserts custom model rules, enabling users to define custom entity types and their detection rules using YAML configuration.

### Changes

**New Resource: `grafana_asserts_custom_model_rules`**
- Manages custom entity definitions through the Asserts CustomModelRulesControllerAPI
- Supports flexible entity type configuration with Prometheus query-based detection
- Implements full CRUD operations with proper retry logic for eventual consistency
- Uses name-based resource identification for individual rule set management

**Key Features**
- Entity type definition with custom names and types (Service, Pod, Namespace, etc.)
- Prometheus query-based entity detection rules
- Nested schema structure supporting multiple entities per rule set
- Error handling and retry mechanisms

### Technical Details

The resource implements proper retry logic using shared utilities to handle eventual consistency issues common with distributed systems. The implementation includes conversion functions between Terraform schema and API DTOs, ensuring proper data transformation and validation.

### Testing

- Unit tests validate all CRUD operations and data conversion
- Acceptance tests verify resource lifecycle and state management
- Stress tests confirm retry logic handles concurrent operations
- Import functionality verified with proper state handling

### Breaking Changes

None. This is a new resource addition with no impact on existing functionality.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/grafana/terraform-provider-grafana?pullRequest=2338) <br/>All committers have signed the CLA.

matt-metivier committed:
4658: docs update

feat(asserts): make custom model rules resource more Terraform native

- Replace YAML string with structured blocks for better UX
- Add conversion functions between Terraform blocks and API DTOs
- Update schema to use nested blocks (rules -> entity -> defined_by)
- Provide better IDE support, validation, and type safety
- Update tests, documentation, and examples to use new format
- Follow Terraform conventions similar to Kubernetes provider

This addresses the PR feedback from Alex3k to make the resource
more idiomatic and user-friendly while maintaining API compatibility.

test: add stress test flag to control eventual consistency tests

- Add CheckStressTestsEnabled function to testutils package
- Update all stress tests to require TF_ACC_STRESS environment variable
- Stress tests will now be skipped unless explicitly enabled
- This prevents flaky stress test failures in CI while still allowing
  developers to run them when needed for testing eventual consistency

feat(asserts): add custom model rules resource

- Add grafana_asserts_custom_model_rules resource for managing Asserts custom model rules
- Implement full CRUD operations with proper retry logic using shared utilities
- Add comprehensive tests including stress tests for eventual consistency
- Add documentation and usage examples
- Update catalog configuration for the new resource
- Register the new resource in the asserts resources list

This resource allows users to define custom entity types and their detection rules
using YAML configuration, enabling flexible monitoring model customization.

Alex3k commented on docs/resources/asserts_custom_model_rules.md:
```diff
@@ -0,0 +1,46 @@
+    - type: Whatever
+      name: Nothing
+      definedBy:
+      - query: "up{job=\\"nothing\\"}"
```
Would it make sense to make this more Terraform native by creating it like 
```
resource "grafana_asserts_custom_model_rules" "test" {
  name  = "test-anything"
  rules {
    entity {
      type = "whatever"
      name = "foo"
      definedBy {
        query = "up{job=\\"nothing\\"}"
      }

      definedBy {
        query = "up{job=\\"nothing\\"}"
      }
    }
  }
}
```

I feel this follows other conventions like in the k8s provider when defining aspects like environment variables as an example. 

Yaml provider a lot of opportunity for error

matt-metivier replied:
Thanks for the suggestion, I updated the string

github-actions[bot] committed:
4658: expanding the tests

github-actions[bot] committed:
4658: expanding the tests

github-actions[bot] committed:
4658: expanding the docs

github-actions[bot] committed:
4658: expanding the converters

github-actions[bot] committed:
4658: fixing test

github-actions[bot] committed:
4658: regenerating docs/fmt

github-actions[bot] committed:
4658: updating go client

matt-metivier committed:
Merge branch 'main' into asserts-custom-model-rules

github-actions[bot] committed:
4658: fixing concurrency test

github-actions[bot] committed:
Merge remote-tracking branch 'origin/asserts-custom-model-rules' into asserts-custom-model-rules

github-actions[bot] committed:
4658: Adding listener

egovova11 commented on docs/resources/asserts_custom_model_rules.md:
```diff
@@ -0,0 +1,108 @@
+
+Optional:
+
+- `disabled` (Boolean) Whether this query is disabled.
```
I think it should say something like `whether is this rule is disabled`. Specifically, when `disabled` flag is set to `true`, then only the `query` field is used to match a rule to disable. 

This may save some time for people sifting though the docs.

matt-metivier replied:
added a better description on it

github-actions[bot] committed:
4658: addressing small things from my review

matt-metivier committed:
Merge branch 'main' into asserts-custom-model-rules

github-actions[bot] committed:
4658: fixing other teams failing test blocking pipeline

github-actions[bot] committed:
Merge remote-tracking branch 'origin/asserts-custom-model-rules' into asserts-custom-model-rules

github-actions[bot] committed:
4658: fixing other teams failing test blocking pipeline

jgehrett commented on docs/resources/asserts_custom_model_rules.md:
```diff
@@ -0,0 +1,108 @@
+Optional:
+
+- `disabled` (Boolean) Whether this entity is disabled.
+- `enriched_by` (List of String) List of enrichment sources for the entity.
```
`enriched_by` has the same schema as `defined_by`.
It's a [`PropertyRule`](https://github.com/grafana/asserts-adi/blob/master/model/rules/src/main/java/ai/asserts/model/rules/PropertyRule.java)

jgehrett commented on internal/resources/asserts/resource_custom_model_rules.go:
```diff
@@ -0,0 +1,478 @@
+									"enriched_by": {
+										Type:        schema.TypeList,
+										Optional:    true,
+										Description: "List of enrichment sources for the entity.",
```
I think this should be similar to `defined_by`.
Something like:
`List of queries that enrich entities of this type.`

jgehrett reviewed (commented):
Looking good! Sorry I'm late to the party. Just a small comment on `enriched_by` doc
